src/java.base/share/classes/jdk/internal/foreign/Utils.java

```diff
@@ -56,28 +59,39 @@ public final class Utils {
     // Suppresses default constructor, ensuring non-instantiability.
     private Utils() {}
 
-    private static final MethodHandle BYTE_TO_BOOL;
-    private static final MethodHandle BOOL_TO_BYTE;
-    private static final MethodHandle ADDRESS_TO_LONG;
-    private static final MethodHandle LONG_TO_ADDRESS_TARGET;
-    private static final MethodHandle LONG_TO_ADDRESS_NO_TARGET;
+    private static final int
+            FILTER_ADDRESS_TO_LONG = 0,
+            FILTER_LONG_TO_POINTER = 1,
+            FILTER_LONG_TO_SEGMENT_WITH_LAYOUT = 2,
+            FILTER_LIMIT = 3;
+    private static final @Stable MethodHandle[] VH_FILTERS = new MethodHandle[FILTER_LIMIT];
+
+    private static MethodHandle filterHandle(int index) {
+        var ret = VH_FILTERS[index];
+        if (ret != null)
+            return ret;
+        return computeFilterHandle(index);
+    }
 
-    static {
-        try {
-            MethodHandles.Lookup lookup = MethodHandles.lookup();
-            BYTE_TO_BOOL = lookup.findStatic(Utils.class, "byteToBoolean",
-                    MethodType.methodType(boolean.class, byte.class));
-            BOOL_TO_BYTE = lookup.findStatic(Utils.class, "booleanToByte",
-                    MethodType.methodType(byte.class, boolean.class));
-            ADDRESS_TO_LONG = lookup.findStatic(SharedUtils.class, "unboxSegment",
-                    MethodType.methodType(long.class, MemorySegment.class));
-            LONG_TO_ADDRESS_TARGET = lookup.findStatic(Utils.class, "longToAddress",
-                    MethodType.methodType(MemorySegment.class, long.class, AddressLayout.class));
-            LONG_TO_ADDRESS_NO_TARGET = lookup.findStatic(Utils.class, "longToAddress",
-                    MethodType.methodType(MemorySegment.class, long.class));
-        } catch (Throwable ex) {
-            throw new ExceptionInInitializerError(ex);
-        }
+    private static MethodHandle computeFilterHandle(int index) {
+        MethodHandles.Lookup lookup = MethodHandles.lookup();
```

Comment:
please break this big conditional up!

Reply:
This is a one-time cold path initializer so I did not care too much about its size. How should I break it up?

Reply:
I mean - using a single expression which is a conditional with two big switches seems quite unreadable.I suggest to move the `VH_FILTERS[index] = ` assignment out of the `computeFilterHandle` method. Then, rewrite `computeFilterHandle` as:

```
if (addressSize != 4) {
    return switch ...
} else {
    return switch ...
}
```